### PR TITLE
feat(TimePicker): class -> functional component

### PR DESCRIPTION
**Pull request by sstrubberg**
REF #9712 

Refactored `TimePicker` into a functional component with RTL tests.

- Removed instances of `evt.persist();` because it is no longer needed as of React 17. However, this causes Storybook to generate console error messages when using Actions, so I removed them.

#### Testing / Reviewing

1. Pull down locally
2. Go to `packages/react`
3. Run `yarn start:v11` to run old storybook with v11 components.
4. Ensure tests for Next versions of `TimePicker` go green.

### CODE REVIEW

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React from 'react';
+import { default as TimePicker } from './TimePicker';
+
+import { render, cleanup, screen, fireEvent } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import '@testing-library/jest-dom';
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js]:
```suggestion
```

This is automatically included in our Jest entrypoint 👍

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+describe('TimePicker', () => {
+  describe('input', () => {
+    afterEach(cleanup);
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js]:
```suggestion
```

This is automatically run by RTL so feel free to remove it 👍

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    it('renders as expected', () => {
+      render(<TimePicker id="time-picker"></TimePicker>);
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js]:
```suggestion
      render(<TimePicker id="time-picker" />);
```

If a component has no children we can use `<ComponentName />` as a shorthand

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      expect(onClick).not.toHaveBeenCalled();
+    });
+
+    it('should set placeholder as expected', () => {
+      render(<TimePicker id="time-picker" placeholder="🧸"></TimePicker>);
+      expect(screen.getByPlaceholderText('🧸')).toBeInTheDocument();
+    });
+  });
+  describe('label', () => {
+    afterEach(cleanup);
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js]:
```suggestion
```

**Review by joshblack — CHANGES_REQUESTED** [review]
Looking great! Just left a couple of style nits

File: packages/react/src/components/TimePicker/next/TimePicker.js [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        onBlur(event);
+      }
+    },
+    pattern,
+    placeholder,
+    maxLength,
+    id,
+    type,
+    disabled,
+    value,
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker.js]: Random refactoring thought, not sure about the content from the original file but can we just put these on the component below directly? 🤔  The one that is the `<input>`

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+      fireEvent.click(screen.getByRole('textbox'));
+      expect(onClick).not.toHaveBeenCalled();
+    });
+
+    it('should set placeholder as expected', () => {
+      render(<TimePicker id="time-picker" placeholder="🧸"></TimePicker>);
+      expect(screen.getByPlaceholderText('🧸')).toBeInTheDocument();
+    });
+  });
+  describe('label', () => {
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js:49]: Style nit for spacing

```suggestion

  describe('label', () => {
```

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+      render(<TimePicker id="time-picker"></TimePicker>);
+      expect(screen.queryByLabelText('🐳')).not.toBeInTheDocument();
+    });
+
+    it('renders a label as expected', () => {
+      render(<TimePicker id="time-picker" labelText="🐳"></TimePicker>);
+      expect(screen.getByLabelText('🐳')).toBeInTheDocument();
+    });
+  });
+  describe('events', () => {
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js:61]:
```suggestion

  describe('events', () => {
```

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+describe('TimePicker', () => {
+  describe('input', () => {
+    it('renders as expected', () => {
+      render(<TimePicker id="time-picker" />);
+      expect(screen.getByRole('textbox')).toBeInTheDocument();
+    });
+
+    it('passes classNames as expected', () => {
+      render(<TimePicker id="time-picker" className="🚀"></TimePicker>);
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js]:
```suggestion
      render(<TimePicker id="time-picker" className="🚀" />);
```

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+      expect(screen.getByRole('textbox')).toHaveClass('🚀');
+    });
+
+    it('should set type as expected', () => {
+      render(<TimePicker id="time-picker"></TimePicker>);
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js]:
```suggestion
      render(<TimePicker id="time-picker" />);
```

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+      expect(screen.getByRole('textbox')).toHaveAttribute('type', 'text');
+    });
+
+    it('should set value as expected', () => {
+      render(<TimePicker id="time-picker" value="🐶"></TimePicker>);
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js]:
```suggestion
      render(<TimePicker id="time-picker" value="🐶" />);
```

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+      expect(screen.getByRole('textbox')).toHaveAttribute('value', '🐶');
+    });
+
+    it('should set disabled as expected', () => {
+      const onClick = jest.fn();
+      render(
+        <TimePicker id="time-picker" onClick={onClick} disabled></TimePicker>
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js]:
```suggestion
        <TimePicker id="time-picker" onClick={onClick} disabled />
```

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+      );
+      fireEvent.click(screen.getByRole('textbox'));
+      expect(onClick).not.toHaveBeenCalled();
+    });
+
+    it('should set placeholder as expected', () => {
+      render(<TimePicker id="time-picker" placeholder="🧸"></TimePicker>);
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js]:
```suggestion
      render(<TimePicker id="time-picker" placeholder="🧸" />);
```

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+      expect(screen.getByPlaceholderText('🧸')).toBeInTheDocument();
+    });
+  });
+  describe('label', () => {
+    it('does not render a label by default', () => {
+      render(<TimePicker id="time-picker"></TimePicker>);
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js]:
```suggestion
      render(<TimePicker id="time-picker" />);
```

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+      expect(screen.queryByLabelText('🐳')).not.toBeInTheDocument();
+    });
+
+    it('renders a label as expected', () => {
+      render(<TimePicker id="time-picker" labelText="🐳"></TimePicker>);
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js]:
```suggestion
      render(<TimePicker id="time-picker" labelText="🐳" />);
```

File: packages/react/src/components/TimePicker/next/TimePicker-test.js [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+      expect(screen.getByLabelText('🐳')).toBeInTheDocument();
+    });
+  });
+  describe('events', () => {
+    it('should write text inside the textbox', () => {
+      render(<TimePicker id="time-picker"></TimePicker>);
```
Comment (joshblack) [inline packages/react/src/components/TimePicker/next/TimePicker-test.js]:
```suggestion
      render(<TimePicker id="time-picker" />);
```

**Review by joshblack — CHANGES_REQUESTED** [review]

**Review by joshblack — APPROVED** [review]

**Review by tay1orjones — COMMENTED** [review]
The storybook actions throw this error:

> Warning: This synthetic event is reused for performance reasons. If you're seeing this, you're accessing the property `buttons` on a released/nullified synthetic event. This is set to null. If you must keep the original synthetic event around, use event.persist(). See https://fb.me/react-event-pooling for more information.

Will users accessing the `onChange`, `onClick`, and `onBlur` get this same error message  if they're using React v16?

The React  [v17 blog post](https://reactjs.org/blog/2020/08/10/react-v17-rc.html#no-event-pooling) has a brief note that `event.persist()` is still available on the React event object, it just doesn't do anything in React v17. 

Should we continue to use `event.persist()` and wait to remove it until (if?) we remove support for React v16?

**Review by tay1orjones — APPROVED** [review]

**Pull request merged**